cli/azd/cmd/login.go

```diff
@@ -57,50 +76,109 @@ func loginCmdDesign(global *internal.GlobalCommandOptions) (*cobra.Command, *log
 }
 
 type loginAction struct {
-	formatter output.Formatter
-	writer    io.Writer
-	console   input.Console
-	azCli     azcli.AzCli
-	flags     loginFlags
+	formatter   output.Formatter
+	writer      io.Writer
+	console     input.Console
+	authManager *auth.Manager
+	flags       loginFlags
 }
 
 func newLoginAction(
 	formatter output.Formatter,
 	writer io.Writer,
-	azcli azcli.AzCli,
+	authManager *auth.Manager,
 	flags loginFlags,
 	console input.Console,
 ) *loginAction {
 	return &loginAction{
-		formatter: formatter,
-		writer:    writer,
-		console:   console,
-		azCli:     azcli,
-		flags:     flags,
+		formatter:   formatter,
+		writer:      writer,
+		console:     console,
+		authManager: authManager,
+		flags:       flags,
 	}
 }
 
-func (la *loginAction) Run(ctx context.Context) error {
-	if err := tools.EnsureInstalled(ctx, la.azCli); err != nil {
-		return err
-	}
+const (
+	// CodespacesEnvVarName is the name of the env variable set when you're in a Github codespace. It's
+	// just set to 'true'.
+	CodespacesEnvVarName = "CODESPACES"
+
+	// RemoteContainersEnvVarName is the name of the env variable set when you're in a remote container. It's
+	// just set to 'true'.
+	RemoteContainersEnvVarName = "REMOTE_CONTAINERS"
+)
 
+func (la *loginAction) Run(ctx context.Context) error {
 	if !la.flags.onlyCheckStatus {
```

Comment:
Consider inverting this if expression for a short circuit return instead of the deep if block nesting.

Reply:
I extracted all this code into a `login()` method on `loginAction` that actually calls the correct `auth.Manager` method to log in based on the flags.